### PR TITLE
Fix for default methods, etc.

### LINKED ISSUE #219: IClass.getAllMethods() ignores default methods
`com.ibm.wala.classLoader.IClass.getAllMethods()` should "compute the methods declared by this class or any of its superclasses," but it does not consider default methods. Since classes inherit default methods from implemented interfaces, I would expect such methods to be returned.

## Input
`java.util.ArrayList`

## Expected Output
Return value should contain `java.util.Collection.stream()`.

## Actual Output
`java.util.Collection.stream()` is not part of the return value.

### PR DESCRIPTION
**Pull request by khatchad**
- Fix #219 
- Test for #219.
- Minor efficiency improvements.

### CODE REVIEW

**Comment by khatchad** [comment]
Hm, I'm not getting this error on my local machine. Could be related to #222.

**Comment by khatchad** [comment]
Here's the output from my own machine:
```
$ mvn clean verify -DskipTests=true -B -q
mkdir -p /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/
/usr/lib/jvm/java-8-openjdk-amd64/bin/javah -classpath "/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/" -d "/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/" com.ibm.wala.cast.ir.translator.NativeBridge
/usr/lib/jvm/java-8-openjdk-amd64/bin/javah -classpath "/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/" -d "/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/" com.ibm.wala.cast.ir.translator.NativeTranslatorToCAst
echo /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
g++ -pthread -g  -I../../../com.ibm.wala.cast/source/c/include/ -I/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/ -I/usr/lib/jvm/java-8-openjdk-amd64/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -DWALA_CLASSPATH="\"/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/\""  -fPIC -o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o -c jni/Java_com_ibm_wala_cast_ir_translator_NativeBridge.cpp
echo /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
g++ -pthread -g  -I../../../com.ibm.wala.cast/source/c/include/ -I/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/ -I/usr/lib/jvm/java-8-openjdk-amd64/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -DWALA_CLASSPATH="\"/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/\""  -fPIC -o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o -c jni/Exceptions.cpp
echo /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
g++ -pthread -g  -I../../../com.ibm.wala.cast/source/c/include/ -I/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/ -I/usr/lib/jvm/java-8-openjdk-amd64/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -DWALA_CLASSPATH="\"/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/\""  -fPIC -o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o -c jni/launch.cpp
echo /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
g++ -pthread -g  -I../../../com.ibm.wala.cast/source/c/include/ -I/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/ -I/usr/lib/jvm/java-8-openjdk-amd64/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -DWALA_CLASSPATH="\"/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/\""  -fPIC -o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o -c jni/CAstWrapper.cpp
g++ -pthread -shared -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x86_64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x86_64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib -ljvm /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o -o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast.so
ar -r /tmp/WALA/com.ibm.wala.cast/target/classes/libcast.a /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Java_com_ibm_wala_cast_ir_translator_NativeBridge.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast/CAstWrapper.o
ar: creating /tmp/WALA/com.ibm.wala.cast/target/classes/libcast.a
/usr/lib/jvm/java-8-openjdk-amd64/bin/javah -classpath "/tmp/WALA/com.ibm.wala.cast.test/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/" -d "/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/" com.ibm.wala.cast.test.TestNativeTranslator
g++ -pthread -g  -I../../../com.ibm.wala.cast/source/c/include/ -I/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/ -I/usr/lib/jvm/java-8-openjdk-amd64/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -DWALA_CLASSPATH="\"/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/\""  -fPIC -o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//smoke.o -c smoke.cpp
g++ -pthread -shared -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x86_64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x86_64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib -ljvm -Wl,-rpath -Wl,/tmp/WALA/com.ibm.wala.cast/target/classes/ /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//Exceptions.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//CAstWrapper.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//smoke.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//Java_com_ibm_wala_cast_ir_translator_NativeBridge.o -o /tmp/WALA/com.ibm.wala.cast.test/target/classes/libxlator_test.so
g++ -pthread -g  -I../../../com.ibm.wala.cast/source/c/include/ -I/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/ -I/usr/lib/jvm/java-8-openjdk-amd64/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -DWALA_CLASSPATH="\"/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/\""  -fPIC -o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//smoke_main.o -c smoke_main.cpp
g++ -pthread -g  -I../../../com.ibm.wala.cast/source/c/include/ -I/tmp/WALA/com.ibm.wala.cast/target/classes/libcast/ -I/usr/lib/jvm/java-8-openjdk-amd64/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -DWALA_CLASSPATH="\"/tmp/WALA/com.ibm.wala.cast/target/classes/:/tmp/WALA/com.ibm.wala.cast/target/classes/\""  -fPIC /tmp/WALA/com.ibm.wala.cast/target/classes//libcast/launch.o /tmp/WALA/com.ibm.wala.cast/target/classes/libcast//smoke_main.o -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x86_64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x86_64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x64/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/x64/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/server -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/server -L/usr/lib/jvm/java-8-openjdk-amd64/jre/lib -Wl,-rpath -Wl,/usr/lib/jvm/java-8-openjdk-amd64/jre/lib -ljvm -o /tmp/WALA/com.ibm.wala.cast.test/target/classes//smoke_main 
929338653:BINARY_EXPR
  "+"
  "1"
  "2"

/tmp/WALA/com.ibm.wala.cast.test/target/classes:/tmp/WALA/com.ibm.wala.cast.test/../com.ibm.wala.cast/target/classes:/tmp/WALA/com.ibm.wala.cast.test/../com.ibm.wala.core/target/classes:/tmp/WALA/com.ibm.wala.cast.test/../com.ibm.wala.util/target/classes
$ echo $?
0
```

**Comment by msridhar** [comment]
@khatchad can you run `git rebase master` on your branch and fix any conflicts that come up?  Then, try running `mvn clean verify -DskipTests=true -B -q` from your repository root directory and see if any errors arise.

**Comment by khatchad** [comment]
@msridhar Sure, this is what I got:

```
$ git log -n 1
commit 33fd2ccb96f12ca7325f977675f19f058584a826
Author: Raffi Khatchadourian <raffi.khatchadourian@hunter.cuny.edu>
Date:   Tue Aug 15 11:58:35 2017 -0400

    Update JDT prefs for Java 1.8.
$ git fetch
$ git rebase master
Current branch master is up to date.
$ mvn clean verify -DskipTests=true -B -q
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:0.23.1:compile (default-compile) on project com.ibm.wala.core: Compilation failure: Compilation failure:
[ERROR] /home/rk1424/Workspaces/Research/WALA/com.ibm.wala.core/src/com/ibm/wala/ipa/slicer/Slicer.java:[315]
[ERROR] this.supergraph = backward ? BackwardsSupergraph.make(forwards) : forwards;
[ERROR] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from BackwardsSupergraph<Statement,PDG<? extends InstanceKey>> to ISupergraph<Statement,PDG<?>>
[ERROR] /home/rk1424/Workspaces/Research/WALA/com.ibm.wala.core/src/com/ibm/wala/demandpa/alg/SimpleDemandPointsTo.java:[104]
[ERROR] return Predicate.filter(dfs, iKeyFilter);
[ERROR] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] Type mismatch: cannot convert from List<Object> to Collection<InstanceKey>
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :com.ibm.wala.core
```

**Comment by msridhar** [comment]
For me, master does not fail with this error.  So it is most likely something in this commit.  I'm guessing it's the version bumps to 1.8.  Maybe get rid of those and see if it builds?  There is still a discussion about going to 1.8 I think.  Probably best to wait for Julian to do it rather than adding that to this PR.

**Comment by khatchad** [comment]
Ah, sorry about that. I was under the impression that the decision was to go forward with Java 1.8.

On Tue, 2017-08-15 at 16:37 +0000, Manu Sridharan wrote:

For me, master does not fail with this error. So it is most likely something in this commit. I'm guessing it's the version bumps to 1.8. Maybe get rid of those and see if it builds? There is still a discussion about going to 1.8 I think. Probably best to wait for Julian to do it rather than adding that to this PR.

**Comment by khatchad** [comment]
@msridhar:

>  I'm guessing it's the version bumps to 1.8. Maybe get rid of those and see if it builds? 

Done. Looks like it's passing now.

File: com.ibm.wala.core/src/com/ibm/wala/analysis/typeInference/TypeInference.java [hunk 1]
```diff
@@ -727,8 +727,8 @@ public void visitStoreIndirect(SSAStoreIndirectInstruction instruction) {
 
     @Override
     public IVariable makeVariable(int valueNumber) {
-      SymbolTable st = ir.getSymbolTable();
       if (doPrimitives) {
+        SymbolTable st = ir.getSymbolTable();
```
Comment (msridhar) [inline com.ibm.wala.core/src/com/ibm/wala/analysis/typeInference/TypeInference.java]: Are these changes to `TypeInference` part of the fix for default methods?  If not, can we pull all changes to `TypeInference` into a separate PR, to keep things clean?

**Comment by liblit** [comment]
I am confused by the proposed commits that affect `ModRef.java`:

* Proposed commit d8b59306b4057a0fb7ff1daa6b41cabb8cbff7fa *reverses* changes I made in 9159ea763645ff6332d73f67cacb6a97ec6dd413.

* Proposed commit ea49d92cc9546b8bfc5962252293c8093bb8511c replicates changes that I already made in fbf9abfa3aac2b334b2ac6e4df60224de5e8023e. I do not even understand why git considers proposed commit ea49d92cc9546b8bfc5962252293c8093bb8511c to be conflict-free, given that the “before” code in that diff does not match the current `master` code.  For example, part of ea49d92cc9546b8bfc5962252293c8093bb8511c is a change of lines 47–48 from

  ```java
  public static ModRef make() {
    return new ModRef();
  ```

  to

  ```java
  public static <T extends InstanceKey> ModRef<T> make() {
    return new ModRef<T>();
  ```

  but in the current `master`, these lines are already

  ```java
  public static <U extends InstanceKey> ModRef<U> make() {
    return new ModRef<U>();
  ```

**Comment by msridhar** [comment]
@liblit my plan was to do a squash and merge on this one, as the desired change is pretty narrow (see the Files Changed tab) and as you point out the commit history is kind of hairy.  If I do a squash and merge I think all the changes here will show up as one commit on master, just including the final changes.

**Comment by liblit** [comment]
There are other redundant changes too. In this pull request, @khatchad proposes to add `results` to the end of `.gitignore` in 0e19be83a82c6ad23b28b75eb00a7101f679aabc. But that exact change was already made as part of @khatchad’s own 97a6496c1f54a20fbc54599f314f0b4d2827d570 commit back in early May.

I must be very confused about the circumstances under which git (or GitHub) considers a pull request to be conflict-free.

**Comment by liblit** [comment]
@msridhar: aah, that “Files changed” view is indeed enlightening. The redundant file churn I was seeing in the individual commits seems to evaporate away when seen through that lens. This gives me a bit more understanding of what git is doing behind the scenes, and why these changes can be treated as conflict-free here. Thank you for steering me in the right direction.

**Comment by khatchad** [comment]
@liblit It sounds like you are looking at the merge commits. @msridhar yes, I created the commits expecting you to squash them.

File: com.ibm.wala.core/src/com/ibm/wala/analysis/typeInference/TypeInference.java (hunk 1 above)
Comment (khatchad) [inline com.ibm.wala.core/src/com/ibm/wala/analysis/typeInference/TypeInference.java]: > Are these changes to `TypeInference` part of the fix for default methods?

No, they were just on my fork.

>  If not, can we pull all changes to TypeInference into a separate PR, to keep things clean?

I'm not quite sure how to "split" pull requests in GitHub. The best I can think of is to revert the performance improvement commits. Once you accept the PR, I'll unrevert them and issue another request.

**Comment by khatchad** [comment]
@liblit Actually, you're right. The commit history on GitHub looks really strange. It will go away with a squash.

File: com.ibm.wala.core/src/com/ibm/wala/analysis/typeInference/TypeInference.java (hunk 1 above)
Comment (khatchad) [inline com.ibm.wala.core/src/com/ibm/wala/analysis/typeInference/TypeInference.java]: @msridhar Done with dccae97de3079f35cfb9f6591f9d24f3931a7a9c and cd013804fc90c9dbb0832684d850d37656db9d26.

**Comment by msridhar** [comment]
Thanks @khatchad!

**Pull request merged**